hw3/02-charts.js

```diff
@@ -29,24 +31,123 @@ const borderColors = [
 // url for the Thrones API
 const url = 'https://thronesapi.com/api/v2/Characters';
 
-const renderChart = () => {
+
+// async function to fetch the data from a given url
+const fetchData = async function fetchDataFromApi(url) {
+  try {
+      const res = await fetch(url);
+      const data = await res.json();
+      return data;
+  }
+  catch {
+      console.error('Error', error);
+  }
+};
+
+// function to count family names
+const getFam = function getFamilyNames(data) {
+
+  let familyNames = [];
+
+  data.forEach(element => {
+    if(!familyNames.some(name => name.name === element.lastName)) {
+      familyNames.push(
+        {
+          'name': element.lastName,
+          'count': 1
+        });
+    }
+    else {
+      let nameIndex = familyNames.findIndex(name => name.name === element.lastName);
+      familyNames[nameIndex].count++;
+    }
+  });
+
+  let filteredNames = [];
+  familyNames.forEach(element => {
+    if(element.name === 'Targaryan')
+    {
+      let tarName = familyNames.findIndex(name => name.name === 'Targaryen');
+      familyNames[tarName].count += element.count;
+    }
+    else if(element.name === '' || element.name === 'None') {
+      let unName = familyNames.findIndex(name => name.name === 'Unknown');
+      familyNames[unName].count += element.count;
+    }
+    else {
+      filteredNames.push(element);
+    }
+  });
+
+  return filteredNames;
+}
+
+// function to render the chart of families
+// returns a list of last names with only one entry
+const renderFamilyChart = (nameData) => {
   const donutChart = document.querySelector('.donut-chart');
 
+  let labelNames = ['No Family Gang :('];
+  let nameCounts = [0];
+  let noFamily = [];
+  nameData.forEach(element => {
+
+    if(element.count > 1){
+      labelNames.push(element.name);
+      nameCounts.push(element.count);  
+    }
+    else {
+      nameCounts[0]++;
+      noFamily.push(element.name);
+    }
+
+  });
+
   new Chart(donutChart, {
     type: 'doughnut',
     data: {
-      labels: ['label', 'label', 'label', 'label'],
+      labels: labelNames,
       datasets: [
         {
           label: 'My First Dataset',
-          data: [1, 12, 33, 5],
+          data: nameCounts,
           backgroundColor: backgroundColors,
           borderColor: borderColors,
           borderWidth: 1,
         },
       ],
     },
   });
+
+  return noFamily;
+};
+
+// function that prints a list of people with no family members
+const noFamList = function displaynoFamily(list) {
+  let header = document.createElement('h2');
+  header.innerText = 'These are the last names that have no family :(';
+  let lackFam = document.createElement('ul');
+  list.forEach(element => {
+    let listItem = document.createElement('li');
```

Comment:
All of these `let` statements (and some others in the file) can be `const`.

`let` is only needed when the variable itself is reassigned (i.e. `header = ...`).